src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -18,23 +18,33 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include <type_traits>
+#include "FrontendDialectTransformer.hpp"
+#include "src/Interface/HasOnnxSubgraphOpInterface.hpp"
+#include "src/Interface/ResultTypeInferenceOpInterface.hpp"
 
 #include "mlir/IR/BuiltinOps.h"
-#include "onnx/defs/schema.h"
 #include "llvm/ADT/TypeSwitch.h"
 
+#if defined(__clang__)
```

Comment:
> I think less duplication of the code to disable/enable the warnings around the onnx headers would be better. Perhaps you can add a macro or two to do this for you and re-use them as necessary.

Sure. That can be done. 

> I would still like to understand what the specific warnings are and disable them judiciously as opposed all of them at once.

Trying to understand this request better.  I am starting from the assumption we want to suppress all warnings in third_party code so that we can easily focus on warnings coming from the onnx-mlir code.  That is why I just added a pragma to suppress all the warnings.  The alternative would be to suppress each warning individually.  This second approach has the disadvantage that it would require more maintenance as third_party projects evolve and we pull new versions in our repo.   A new version might introduce a new warning, which would require someone to add another pragma to suppress it.  Because of the added maintenance cost IMO is better to disable them all.   It is easy enough to find out what warnings the third_party code might have by writing a small test code that includes the header.

Reply:
I think I mentioned this elsewhere, but what the warnings are is important. Disabling all warnings in submodules is definitely easier, however, some of the warnings in the submodules could be indicative of bugs that can impact onnx-mlir. So by disabling all warnings we may be putting ourselves in a position where we have to spend a lot of time investigating a bug that could have easily been found if the warnings were not disabled.

My suggestion is to look at what the current warnings are and see what/how many buckets they fall in. I suspect there are only a handful of types of warnings and it will be possible to have a look at the list and decide that they are just not things we are concerned about.

Bug as @gongsu832 said, we might want to fix some of the warnings directly in the third party code instead of disabling them here, because they could actually be bugs that need to be addressed. We are all working on various parts of this project and sometimes in multiple of these projects, so to some extent we're collectively responsible for making sure that code quality across all of the dependencies improves. That doesn't mean we should go fix all the warnings in onnx, but if there are a couple critical ones that would impact onnx-mlir, they are something we should definitely consider fixing.

Reply:
> I think I mentioned this elsewhere, but what the warnings are is important. Disabling all warnings in submodules is definitely easier, however, some of the warnings in the submodules could be indicative of bugs that can impact onnx-mlir. So by disabling all warnings we may be putting ourselves in a position where we have to spend a lot of time investigating a bug that could have easily been found if the warnings were not disabled.

> My suggestion is to look at what the current warnings are and see what/how many buckets they fall in. I suspect there are only a handful of types of warnings and it will be possible to have a look at the list and decide that they are just not things we are concerned about.

OK, I understand your concern better.  Here are the warnings **currently** coming from the header files included by `src/Builder/FrontendDialectTransformer.cpp`. 

```
[build] third_party/onnx/onnx/defs/schema.h:53:25: warning: 'getAttribute' overrides a member function but is not marked 'override' [-Wsuggest-override]
[build]   const AttributeProto* getAttribute(const std::string& name) const {
[build]                         ^
[build] third_party/onnx/onnx/defs/schema.h:31:33: note: overridden virtual function is here
[build]   virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
[build]                                 ^
[build] third_party/onnx/onnx/defs/schema.h:62:8: warning: 'hasInput' overrides a member function but is not marked 'override' [-Wsuggest-override]
[build]   bool hasInput(int inputIndex) const {
[build]        ^
[build] third_party/onnx/onnx/defs/schema.h:32:16: note: overridden virtual function is here
[build]   virtual bool hasInput(int inputIndex) const = 0;
[build]                ^
[build] third_party/onnx/onnx/defs/schema.h:68:8: warning: 'hasOutput' overrides a member function but is not marked 'override' [-Wsuggest-override]
[build]   bool hasOutput(int inputIndex) const {
[build]        ^
[build] third_party/onnx/onnx/defs/schema.h:33:16: note: overridden virtual function is here
[build]   virtual bool hasOutput(int inputIndex) const = 0;
[build]                ^
[build] third_party/onnx/onnx/defs/schema.h:74:20: warning: 'getInputType' overrides a member function but is not marked 'override' [-Wsuggest-override]
[build]   const TypeProto* getInputType(int inputIndex) const {
[build]                    ^
[build] third_party/onnx/onnx/defs/schema.h:36:28: note: overridden virtual function is here
[build]   virtual const TypeProto* getInputType(int inputIndex) const = 0;
[build]                            ^
[build] third_party/onnx/onnx/common/tensor.h:285:1: warning: cast from 'const char *' to 'float *' drops const qualifier [-Wcast-qual]
[build] define_data(float, float_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:285:1: warning: cast from 'const char *' to 'float *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:286:1: warning: cast from 'const char *' to 'double *' drops const qualifier [-Wcast-qual]
[build] define_data(double, double_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:286:1: warning: cast from 'const char *' to 'double *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:287:1: warning: cast from 'const char *' to 'int *' drops const qualifier [-Wcast-qual]
[build] define_data(int32_t, int32_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:287:1: warning: cast from 'const char *' to 'int *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:288:1: warning: cast from 'const char *' to 'long long *' drops const qualifier [-Wcast-qual]
[build] define_data(int64_t, int64_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:288:1: warning: cast from 'const char *' to 'long long *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:289:1: warning: cast from 'const char *' to 'unsigned long long *' drops const qualifier [-Wcast-qual]
[build] define_data(uint64_t, uint64_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:289:1: warning: cast from 'const char *' to 'unsigned long long *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:290:1: warning: cast from 'const char *' to 'std::__1::basic_string<char> *' drops const qualifier [-Wcast-qual]
[build] define_data(std::string, string_data_);
[build] ^
[build] third_party/onnx/onnx/common/tensor.h:270:21: note: expanded from macro 'define_data'
[build]       return (type*)&raw_data_.data()[0];         \
[build]                     ^
[build] third_party/onnx/onnx/common/tensor.h:290:1: warning: cast from 'const char *' to 'std::__1::basic_string<char> *' drops const qualifier [-Wcast-qual]
[build] third_party/onnx/onnx/common/tensor.h:279:21: note: expanded from macro 'define_data'
[build]       return (type*)(raw_data_.data());           \
[build]                     ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:14:
[build] third_party/onnx/onnx/version_converter/adapters/axes_input_to_attribute.h:38:34: warning: cast from 'const char *' to 'long long *' drops const qualifier [-Wcast-qual]
[build]         int64_t* raw = (int64_t*)raw_data.c_str();
[build]                                  ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:17:
[build] third_party/onnx/onnx/version_converter/adapters/batch_normalization_8_9.h:26:21: warning: implicit conversion turns string literal into bool: 'const char [91]' to 'bool' [-Wstring-conversion]
[build]         ONNX_ASSERT("Unsupported conversion due to change in dimensions of inputs when spatial is set to false.");
[build]         ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] third_party/onnx/onnx/common/assertions.h:40:22: note: expanded from macro 'ONNX_ASSERT'
[build]   if (_ONNX_EXPECT(!(cond), 0)) {                   \
[build]                    ~ ^~~~
[build] third_party/onnx/onnx/common/assertions.h:34:47: note: expanded from macro '_ONNX_EXPECT'
[build] #define _ONNX_EXPECT(x, y) (__builtin_expect((x), (y)))
[build]                                               ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:20:
[build] third_party/onnx/onnx/version_converter/adapters/cast_9_8.h:20:25: warning: implicit conversion turns string literal into bool: 'const char [50]' to 'bool' [-Wstring-conversion]
[build]             ONNX_ASSERT("Casting From/To STRING data type is not supported")
[build]             ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] third_party/onnx/onnx/common/assertions.h:40:22: note: expanded from macro 'ONNX_ASSERT'
[build]   if (_ONNX_EXPECT(!(cond), 0)) {                   \
[build]                    ~ ^~~~
[build] third_party/onnx/onnx/common/assertions.h:34:47: note: expanded from macro '_ONNX_EXPECT'
[build] #define _ONNX_EXPECT(x, y) (__builtin_expect((x), (y)))
[build]                                               ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:33:
[build] third_party/onnx/onnx/version_converter/adapters/reshape_5_4.h:33:37: warning: cast from 'const char *' to 'long long *' drops const qualifier [-Wcast-qual]
[build]           int64_t* raw = (int64_t*) raw_data.c_str();
[build]                                     ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:34:
[build] third_party/onnx/onnx/version_converter/adapters/scan_8_9.h:36:23: warning: implicit conversion turns string literal into bool: 'const char [34]' to 'bool' [-Wstring-conversion]
[build]           ONNX_ASSERT("Unsupported conversion to opset 9");
[build]           ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] third_party/onnx/onnx/common/assertions.h:40:22: note: expanded from macro 'ONNX_ASSERT'
[build]   if (_ONNX_EXPECT(!(cond), 0)) {                   \
[build]                    ~ ^~~~
[build] third_party/onnx/onnx/common/assertions.h:34:47: note: expanded from macro '_ONNX_EXPECT'
[build] #define _ONNX_EXPECT(x, y) (__builtin_expect((x), (y)))
[build]                                               ^
[build] In file included from src/Builder/FrontendDialectTransformer.cpp:29:
[build] In file included from third_party/onnx/onnx/version_converter/convert.h:38:
[build] third_party/onnx/onnx/version_converter/adapters/split_13_12.h:34:34: warning: cast from 'const char *' to 'long long *' drops const qualifier [-Wcast-qual]
[build]         int64_t* raw = (int64_t*)raw_data.c_str();
[build]                                  ^
[build] 22 warnings generated
```

Of course the same set of 22 warning messages would reappear every time the same external header is in any of our source files. 

 
> Bug as @gongsu832 said, we might want to fix some of the warnings directly in the third party code instead of disabling them here, because they could actually be bugs that need to be addressed. We are all working on various parts of this project and sometimes in multiple of these projects, so to some extent we're collectively responsible for making sure that code quality across all of the dependencies improves. That doesn't mean we should go fix all the warnings in onnx, but if there are a couple critical ones that would impact onnx-mlir, they are something we should definitely consider fixing.

Agree. I do not actually see suppressing the warnings here as preventing anyone from addressing the problem in one of the other projects.  However I submit is important to reduce noise in other projects now, so that practically it will be  easier to pay attention to warnings in the `onnx-mlir` code (given that is our direct responsibility). Also suppressing warnings from other projects would allow us to instruct the build compiler(s) to turn those remaining warnings in the project to errors. That will have the effect of preventing the introduction of further warnings in future PRs which I do see as good practice. I think we can all agree to that.

Reply:
I've been thinking about how to proceed @sstamenova. If we were to selectively disable only the warning categories that cause issues (namely  -Wsuggest-override, -Wstring-conversion, -Wcast-qual) and add a macro to avoid repeating code (as suggested), would that be agreeable ?  Trying to find a middle ground.  If third_part projects expose more warning categories we can add to the macro, is a bit more maintenance but hopefully  that will not happen often enough to become a problem. What do you think?